lib/src/embedded/dispatcher.dart

```diff
@@ -30,83 +30,65 @@ final _outboundRequestId = 0;
+
+  /// The sink for sending messages to the host.
+  final StreamSink<Uint8List> _sink;
 
   /// The compilation ID for which this dispatcher is running.
   ///
   /// This is added to outgoing messages but is _not_ parsed from incoming
   /// messages, since that's already handled by the [IsolateDispatcher].
-  final int _compilationId;
+  int _compilationId = 0;
 
   /// [_compilationId], serialized as a varint.
-  final Uint8List _compilationIdVarint;
-
-  /// Whether this dispatcher has received its compile request.
-  var _compiling = false;
+  Uint8List _compilationIdVarint = Uint8List.fromList([0]);
 
-  /// A completer awaiting a response to an outbound request.
-  ///
-  /// Since each [Dispatcher] is only running a single-threaded compilation, it
-  /// can only ever have one request outstanding.
-  Completer<GeneratedMessage>? _outstandingRequest;
+  /// Whether a fatal error has occured during host request.
+  var _asyncError = false;
 
   /// Creates a [Dispatcher] that sends and receives encoded protocol buffers
   /// over [channel].
-  Dispatcher(this._channel, this._compilationId)
-      : _compilationIdVarint = serializeVarint(_compilationId);
+  Dispatcher(this._mailbox, this._sink);
 
   /// Listens for incoming `CompileRequests` and runs their compilations.
-  ///
-  /// This may only be called once. Returns whether or not the compilation
-  /// succeeded.
-  Future<bool> listen() async {
-    var success = false;
-    await _channel.stream.listen((binaryMessage) async {
-      // Wait a single microtask tick so that we're running in a separate
-      // microtask from the initial request dispatch. Otherwise, [waitFor] will
-      // deadlock the event loop fiber that would otherwise be checking stdin
-      // for new input.
-      await Future<void>.value();
+  void listen() {
+    do {
+      var packet = _mailbox.take();
+      if (packet.isEmpty) break;
 
       try {
-        InboundMessage? message;
+        var (compilationId, messageBuffer) = parsePacket(packet);
+
+        _compilationId = compilationId;
+        _compilationIdVarint = serializeVarint(compilationId);
+
+        InboundMessage message;
         try {
-          message = InboundMessage.fromBuffer(binaryMessage);
+          message = InboundMessage.fromBuffer(messageBuffer);
         } on InvalidProtocolBufferException catch (error) {
           throw parseError(error.message);
         }
 
         switch (message.whichMessage()) {
-          case InboundMessage_Message.versionRequest:
-            throw paramsError("VersionRequest must have compilation ID 0.");
-
           case InboundMessage_Message.compileRequest:
-            if (_compiling) {
-              throw paramsError(
-                  "A CompileRequest with compilation ID $_compilationId is "
-                  "already active.");
+            var request = message.compileRequest;
+            var response = _compile(request);
+            if (!_asyncError) {
```

Comment:
This is because we want to capture the `ProtocolError` thrown when parsing the host response. `_compile` has the issue where it will capture the `ProtocolError` and convert it to `SassException`. So this check is to prevent it from sending a `CompileFailure` after sending a `ProtocolError`.

I also thought about let `_compile` simply throw `ProtocolError` but that would require change to the non-embedded part of the code base, and introduce `ProtocolError` to the general error handling code.

Reply:
We can try let ProtocolError throw all the way through here: https://github.com/sass/dart-sass/blob/58cbab42e63500f4269bc3f5ed01f7ee311cc4d3/lib/src/visitor/evaluate.dart#L1668-L1675

However, this will introduce ProtocolError to the JS side of the code base, which I'm not sure how to handle in a clean way. Another option is to capture ProtocolError in the `HostCallable` and re-throw it as a different error type e.g. FatalError, and let that particular error fall-through the catch statement all the way.

Reply:
I just pushed a new commit implementing the idea described above.

Reply:
Reverted the commit and update the name and documentation accordingly.